// angular-code-input/esm2015/lib/code-input.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/code-input.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CodeInputComponent } from './code-input.component';
import { CommonModule } from '@angular/common';
import { CodeInputComponentConfigToken } from './code-input.component.config';
export class CodeInputModule {
    /**
     * @param {?} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: CodeInputModule,
            providers: [
                { provide: CodeInputComponentConfigToken, useValue: config }
            ]
        };
    }
}
CodeInputModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    CodeInputComponent
                ],
                exports: [
                    CodeInputComponent
                ]
            },] }
];
