// angular-code-input/esm2015/lib/code-input.component.config.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/code-input.component.config.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export const CodeInputComponentConfigToken = new InjectionToken('CodeInputComponentConfig');
/**
 * @record
 */
export function CodeInputComponentConfig() { }
if (false) {
    /** @type {?|undefined} */
    CodeInputComponentConfig.prototype.codeLength;
    /** @type {?|undefined} */
    CodeInputComponentConfig.prototype.inputType;
    /** @type {?|undefined} */
    CodeInputComponentConfig.prototype.initialFocusField;
    /** @type {?|undefined} */
    CodeInputComponentConfig.prototype.isCharsCode;
    /** @type {?|undefined} */
    CodeInputComponentConfig.prototype.isCodeHidden;
    /** @type {?|undefined} */
    CodeInputComponentConfig.prototype.isPrevFocusableAfterClearing;
    /** @type {?|undefined} */
    CodeInputComponentConfig.prototype.isFocusingOnLastByClickIfFilled;
    /** @type {?|undefined} */
    CodeInputComponentConfig.prototype.code;
}
/** @type {?} */
export const defaultComponentConfig = {
    codeLength: 4,
    inputType: 'tel',
    initialFocusField: undefined,
    isCharsCode: false,
    isCodeHidden: false,
    isPrevFocusableAfterClearing: true,
    isFocusingOnLastByClickIfFilled: false,
    code: undefined
};
